src/backend/avx2/edwards.rs

```diff
@@ -0,0 +1,1057 @@
+
+    fn neg(self) -> CachedPoint {
+        let mut neg = *self;
+        neg.0.swap_AB();
+        neg.0.negate_lazy(D_LANES);
+        neg
+    }
+}
+
+impl<'a> Neg for &'a ExtendedPoint {
+    type Output = ExtendedPoint;
+
+    fn neg(self) -> ExtendedPoint {
+        let mut neg = *self;
+        // (X Y Z T) -> (-X Y Z -T)
+        neg.0.negate(A_LANES | D_LANES);
+        neg
+    }
+}
+
+impl ExtendedPoint {
+    fn double(&self) -> ExtendedPoint {
+        unsafe {
+            use stdsimd::vendor::_mm256_permute2x128_si256;
+            use stdsimd::vendor::_mm256_permutevar8x32_epi32;
+            use stdsimd::vendor::_mm256_blend_epi32;
+            use stdsimd::vendor::_mm256_shuffle_epi32;
+
+            let P = &self.0;
+
+            let mut t0 = FieldElement32x4::zero();
+            let mut t1 = FieldElement32x4::zero();
+
+            // Want to compute (X1 Y1 Z1 X1+Y1).
+            // Not sure how to do this less expensively than computing
+            // (X1 Y1 Z1 T1) --(256bit shuffle)--> (X1 Y1 X1 Y1)
+            // (X1 Y1 X1 Y1) --(2x128b shuffle)--> (Y1 X1 Y1 X1)
+            // and then adding.
+
+            // Set t0 = (X1 Y1 X1 Y1)
+            t0.0[0] = _mm256_permute2x128_si256(P.0[0].into(), P.0[0].into(), 0b0000_0000).into();
+            t0.0[1] = _mm256_permute2x128_si256(P.0[1].into(), P.0[1].into(), 0b0000_0000).into();
+            t0.0[2] = _mm256_permute2x128_si256(P.0[2].into(), P.0[2].into(), 0b0000_0000).into();
+            t0.0[3] = _mm256_permute2x128_si256(P.0[3].into(), P.0[3].into(), 0b0000_0000).into();
+            t0.0[4] = _mm256_permute2x128_si256(P.0[4].into(), P.0[4].into(), 0b0000_0000).into();
+
+            // Set t1 = (Y1 X1 Y1 X1)
+            t1.0[0] = _mm256_shuffle_epi32(t0.0[0].into(), 0b10_11_00_01).into();
+            t1.0[1] = _mm256_shuffle_epi32(t0.0[1].into(), 0b10_11_00_01).into();
+            t1.0[2] = _mm256_shuffle_epi32(t0.0[2].into(), 0b10_11_00_01).into();
+            t1.0[3] = _mm256_shuffle_epi32(t0.0[3].into(), 0b10_11_00_01).into();
+            t1.0[4] = _mm256_shuffle_epi32(t0.0[4].into(), 0b10_11_00_01).into();
+
+            // Set t0 = (X1+Y1 X1+Y1 X1+Y1 X1+Y1)
+            t0.0[0] = t0.0[0] + t1.0[0];
+            t0.0[1] = t0.0[1] + t1.0[1];
+            t0.0[2] = t0.0[2] + t1.0[2];
+            t0.0[3] = t0.0[3] + t1.0[3];
+            t0.0[4] = t0.0[4] + t1.0[4];
+
+            // Set t0 = (X1 Y1 Z1 X1+Y1)
+            // why does this intrinsic take an i32 for the imm8 ???
+            t0.0[0] = _mm256_blend_epi32(P.0[0].into(), t0.0[0].into(), D_LANES as i32).into();
+            t0.0[1] = _mm256_blend_epi32(P.0[1].into(), t0.0[1].into(), D_LANES as i32).into();
+            t0.0[2] = _mm256_blend_epi32(P.0[2].into(), t0.0[2].into(), D_LANES as i32).into();
+            t0.0[3] = _mm256_blend_epi32(P.0[3].into(), t0.0[3].into(), D_LANES as i32).into();
+            t0.0[4] = _mm256_blend_epi32(P.0[4].into(), t0.0[4].into(), D_LANES as i32).into();
```

Comment:
What are we `into()`ing into? And why are we doing `as` type conversions every single time?

Reply:
The problem is that the `_mm256_blend_epi32` intrinsic is defined to take an `i32x8` instead of a `u32x8`, even though the actual instruction works on any 256-bit integer data, shuffling 32-bit lanes according to the immediate. All of these casts disappear from the generated code. It would be nice to file bugs in `stdsimd` to try to make them go away.